docs/authentication.md

```diff
@@ -0,0 +1,130 @@
+# A Guide on ScalarDL Authentication
+
+This document explains the ScalarDL authentication mechanism and how to use it properly. 
+
+## Authentication in ScalarDL
+
+Authentication is one of the key roles in ScalarDL to make the protocol work as expected.
+ScalarDL uses authentication for the following three places:
+* Client authentication (for Ledger and Auditor)
+    * Ledger and Auditor authenticates clients using client-generated signatures attached to requests from the clients.
+* Ledger authentication (for Auditor)
+    * Auditor authenticates Ledger using Ledger-generated signatures attached to [Asset Proofs](how-to-use-proof.md).
+* Auditor authentication (for Ledger)
+    * Ledger authenticates Auditor using Auditor-generated signatures attached to client requests.
+
+Note that Ledger authentication and Auditor authentication are only used in the Auditor mode. For more details about Auditor, please see the [Getting Started with ScalarDL Auditor](getting-started-auditor.md) and [ScalarDL Implementation](implementation.md).
+
+Also, note that we use the term `signature` here to specify a byte array used for authentication.
+
+## Authentication methods
+
+ScalarDL supports two authentication methods: Digital Signatures and HMAC.
+The methods have advantages and disadvantages, as shown below, but Byzantine fault detection capability won't be sacrificed either way.
+
+### Digital Signatures
+* Advantages
+  * Asset records and asset proofs have non-repudiation property. Specifically, a digital signature attached to a request is stored with the corresponding asset records produced by the request so that the records have non-repudiation property, i.e., we can ensure that the records are created by the owner of the private key that signed the request. Moreover, digital signatures attached to asset proofs returned to a client as the result of execution ensure that the proofs are created by Ledger and Auditor, respectively. If a client (application) keeps the proofs, the client can verify results with the proofs as necessary.
+* Disadvantages
+  * Digital Signatures are very slow. They will add non-negligible performance overhead in exchange for the above benefits.
+
+### HMAC
+* Advantages
+    * HMAC is way faster than digital signatures.
+* Disadvantages
+    * Asset records and asset proofs do not have non-repudiation property.
```

Comment:
When reading this part, I just wondered if client requests have the nonrepudiation property. For example, we can say only this client (among all the **clients**) can send this request since only this client knows the secret.
But we cannot distinguish the client and the server since both have the same secret, right? (Can the server impersonate a client and send arbitrary requests to the server itself?)

Reply:
Good point. Yes, client requests also have the nonrepudiation property.
In the HMAC case, as you said, servers can send arbitrary (garbage) requests as a fake client.

Should we just say, we have the nonrepudiation property in client requests in the DS case? Is it good enough?
What do you think?

Reply:
Thanks for the clarification. IMHO, it would be nice if the nonrepudiation property of requests would explicitly be mentioned in both DS and HMAC cases.

Reply:
Thank you for the discussion!
Addressed in [977618f](https://github.com/scalar-labs/scalardl/pull/45/commits/977618f14062db67dc47b0f78b0b2355de56569d).
(Sorry the changes include fixes for other feedback.)